File: udc/udc-serv/src/main/java/com/paypal/udc/service/impl/UserService.java
/*
 * Copyright 2019 PayPal Inc.
 *
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.paypal.udc.service.impl;

import java.io.IOException;
import java.sql.Timestamp;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutionException;
import javax.validation.ConstraintViolationException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.data.elasticsearch.core.ElasticsearchTemplate;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.TransactionSystemException;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;
import com.paypal.udc.dao.user.UserRepository;
import com.paypal.udc.entity.user.User;
import com.paypal.udc.exception.ValidationError;
import com.paypal.udc.service.IUserService;
import com.paypal.udc.util.UserUtil;
import com.paypal.udc.util.enumeration.ActiveEnumeration;


@Service
@Transactional(propagation = Propagation.SUPPORTS, readOnly = true)
public class UserService implements IUserService {

    @Autowired
    private UserRepository userRepository;
    @Autowired
    private UserUtil userUtil;

    @Value("${elasticsearch.user.name}")
    private String esUserIndex;

    @Value("${elasticsearch.type.name}")
    private String esType;

    @Value("${udc.es.write.enabled}")
    private String isEsWriteEnabled;

    @Autowired
    private ElasticsearchTemplate esTemplate;

    private static final SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH-mm-ss");
    final static Logger logger = LoggerFactory.getLogger(UserService.class);

    @Override
    public User getUserById(final long userId) throws ValidationError {
        final User user = this.userUtil.validateUser(userId);
        return user;
    }

    @Override
    public List<User> getAllUsers() {
        final List<User> users = new ArrayList<User>();
        this.userRepository.findAll().forEach(user -> {
            users.add(user);
        });
        return users;
    }

    @Override
    @Transactional(propagation = Propagation.REQUIRED, readOnly = false, rollbackFor = { ValidationError.class,
            ConstraintViolationException.class, DataIntegrityViolationException.class,
            TransactionSystemException.class, IOException.class, InterruptedException.class, ExecutionException.class })
    public User deleteUser(final long userId)
            throws ValidationError, IOException, InterruptedException, ExecutionException {
        final Timestamp timestamp = new Timestamp(System.currentTimeMillis());
        User user = this.userUtil.validateUser(userId);
        user.setUpdatedTimestamp(sdf.format(timestamp));
        user.setIsActiveYN(ActiveEnumeration.NO.getFlag());
        user = this.userRepository.save(user);
        if (this.isEsWriteEnabled.equals("true")) {
            this.userUtil.upsertUsers(this.esUserIndex, this.esType, user, this.esTemplate);
        }
        return user;

    }

    @Override
    @Transactional(propagation = Propagation.REQUIRED, readOnly = false, rollbackFor = { ValidationError.class,
            ConstraintViolationException.class, DataIntegrityViolationException.class,
            TransactionSystemException.class, IOException.class, InterruptedException.class, ExecutionException.class })
    public User addUser(final User user) throws ValidationError, IOException, InterruptedException, ExecutionException {
        final Timestamp timestamp = new Timestamp(System.currentTimeMillis());
        final ValidationError v = new ValidationError();
        final User insertedUser;
        try {
            user.setCreatedUser(user.getUserName());
            user.setUpdatedUser(user.getUserName());
            user.setCreatedTimestamp(sdf.format(timestamp));
            user.setUpdatedTimestamp(sdf.format(timestamp));
            user.setIsActiveYN(ActiveEnumeration.YES.getFlag());
            insertedUser = this.userRepository.save(user);

        }
        catch (final ConstraintViolationException e) {
            v.setErrorCode(HttpStatus.BAD_REQUEST);
            v.setErrorDescription("User name is empty");
            throw v;
        }
        catch (final DataIntegrityViolationException e) {
            v.setErrorCode(HttpStatus.CONFLICT);
            v.setErrorDescription("User name is duplicated");
            throw v;
        }
        if (this.isEsWriteEnabled.equals("true")) {
            this.userUtil.upsertUsers(this.esUserIndex, this.esType, user, this.esTemplate);
        }
        return insertedUser;
    }

    @Override
    @Transactional(propagation = Propagation.REQUIRED, readOnly = false, rollbackFor = { ValidationError.class,
            ConstraintViolationException.class, DataIntegrityViolationException.class,
            TransactionSystemException.class, IOException.class, InterruptedException.class, ExecutionException.class })
    public User updateUser(final User retrievedUser, final User singleSignOnUser)
            throws ValidationError, IOException, InterruptedException, ExecutionException {
        final Timestamp timestamp = new Timestamp(System.currentTimeMillis());
        final ValidationError v = new ValidationError();
        try {
            retrievedUser.setUserFullName(singleSignOnUser.getUserFullName());
            retrievedUser.setManagerName(singleSignOnUser.getManagerName());
            retrievedUser.setQid(singleSignOnUser.getQid());
            retrievedUser.setRoles(singleSignOnUser.getRoles());
            retrievedUser.setOrganization(singleSignOnUser.getOrganization());
            retrievedUser.setLocation(singleSignOnUser.getLocation());
            retrievedUser.setUpdatedTimestamp(sdf.format(timestamp));
            retrievedUser.setCreatedUser(singleSignOnUser.getUserName());
            retrievedUser.setUpdatedUser(singleSignOnUser.getUserName());
            this.userRepository.save(retrievedUser);
        }
        catch (final TransactionSystemException e) {
            v.setErrorCode(HttpStatus.BAD_REQUEST);
            v.setErrorDescription("User name is empty");
            throw v;
        }
        catch (final DataIntegrityViolationException e) {
            v.setErrorCode(HttpStatus.CONFLICT);
            v.setErrorDescription("User name is duplicated");
            throw v;
        }
        if (this.isEsWriteEnabled.equals("true")) {
            this.userUtil.upsertUsers(this.esUserIndex, this.esType, retrievedUser, this.esTemplate);
        }
        return retrievedUser;
    }

    @Override
    public User getUserByName(final String userName) {
        return this.userRepository.findByUserName(userName);
    }

    @Override
    @Transactional(propagation = Propagation.REQUIRED, readOnly = false, rollbackFor = { ValidationError.class,
            ConstraintViolationException.class, DataIntegrityViolationException.class,
            TransactionSystemException.class, IOException.class, InterruptedException.class, ExecutionException.class })
    public User enableUser(final long userId)
            throws ValidationError, IOException, InterruptedException, ExecutionException {
        final Timestamp timestamp = new Timestamp(System.currentTimeMillis());
        User user = this.userUtil.validateUser(userId);
        user.setUpdatedTimestamp(sdf.format(timestamp));
        user.setIsActiveYN(ActiveEnumeration.YES.getFlag());
        user = this.userRepository.save(user);
        if (this.isEsWriteEnabled.equals("true")) {
            this.userUtil.upsertUsers(this.esUserIndex, this.esType, user, this.esTemplate);
        }
        return user;

    }
}

File: udc/udc-web/src/app/udc/catalog/services/user.service.ts
/*
 * Copyright 2019 PayPal Inc.
 *
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {Injectable} from '@angular/core';
import {ApiService} from '../../../core/services/api.service';
import {forkJoin} from 'rxjs/observable/forkJoin';
import {ConfigService} from '../../../core/services/config.service';
import {SessionService} from '../../../core/services/session.service';
import {CatalogService} from './catalog.service';

@Injectable()
export class UserService {

  defaultUserName: string;
  defaultRoleName: string;
  defaultClassificationAdminRoleName: string;
  userFromSessionURL: string;
  udcAdminRole: string;
  udcClassificationAdminRole: string;
  defaultQid: string;
  defaultLocation: string;
  notificationsURL: string;


  constructor(private api: ApiService, private config: ConfigService, private sessionService: SessionService) {
    this.udcAdminRole = 'UDC_ADMIN_ROLE';
    this.udcClassificationAdminRole = 'UDC_CLASSIFICATION_ADMIN';
    this.defaultUserName = 'udcdev';
    this.defaultRoleName = 'UDC_ADMIN_ROLE';
    this.defaultClassificationAdminRoleName = 'UDC_CLASSIFICATION_ADMIN';
    this.defaultQid = 'UDC';
    this.defaultLocation = 'LOCATION';
    this.notificationsURL = api.serverWithPort + '/notification/notifications';
    if (this.api.currentEnv !== 'dev' && this.api.currentEnv !== 'prod') {
      this.userFromSessionURL = api.serverWithPort + '/user/userDetails';
      const user = this.getUserFromSession();
      const notifications = this.getNotificationList();
      forkJoin([user, notifications]).subscribe(results => {
        const userDetails = results[0];
        const notificationList = results[1].filter(notification => notification.activeYN === true);
        this.config.setUserName(userDetails['userName']);
        this.config.setRole(userDetails['roles']);
        const adminStatus: boolean = this.getAdmin(userDetails['roles']);
        this.config.setAdmin(adminStatus);
        const classificationAdminStatus: boolean = this.getClassificationAdmin(userDetails['roles']);
        this.config.setClassificationAdmin(classificationAdminStatus);
        this.config.setQid(userDetails['qid']);
        this.config.setLocation(userDetails['location']);
        this.config.setNotificationCount(notificationList.length);
      }, error => {
      });
    } else {
      this.config.setRole(this.defaultRoleName);
      this.config.setUserName(this.defaultUserName);
      const adminStatus: boolean = this.getAdmin(this.defaultRoleName);
      this.config.setAdmin(adminStatus);
      const classificationAdminStatus: boolean = this.getClassificationAdmin(this.defaultClassificationAdminRoleName);
      this.config.setClassificationAdmin(classificationAdminStatus);
      this.config.setQid(this.defaultQid);
      this.config.setLocation(this.defaultLocation);
      const notifications = this.getNotificationList();
      forkJoin([notifications]).subscribe(results => {
        const notificationList = results[0];
        this.config.setNotificationCount(notificationList.length);
      }, error => {
      });
    }
  }

  getUserFromSession() {
    return this.api.get<Array<any>>(this.userFromSessionURL);
  }

  getNotificationList() {
    return this.api.get<Array<any>>(this.notificationsURL);
  }

  getAdmin(role: string) {
    if (role.search(this.udcAdminRole) === -1) {
      return false;
    } else {
      return true;
    }
  }

  getClassificationAdmin(role: string) {
    if (role.search(this.udcClassificationAdminRole) === -1) {
      return false;
    } else {
      return true;
    }
  }
}
